File: virtool/references.py
import gzip
import json
from cerberus import Validator

import virtool.otus

RIGHTS = [
    "build",
    "modify",
    "modify_otu",
    "remove"
]


def check_import_data(import_data, strict=True, verify=True):
    errors = detect_duplicates(import_data["otus"])

    v = Validator(get_import_schema(require_meta=strict), allow_unknown=True)

    v.validate(import_data)

    if v.errors:
        errors.append({
            "id": "file",
            "issues": v.errors
        })

    otus = dict()

    for otu in import_data["otus"]:
        verification = None

        if verify:
            verification = virtool.otus.verify(otu)

        validation = validate_otu(otu, strict)

        issues = dict()

        if verification:
            issues["verification"] = verification

        if validation:
            issues["validation"] = validation

        if issues:
            otus[otu["_id"]] = issues

    return errors


def detect_duplicate_abbreviation(joined, duplicates, seen):
    abbreviation = joined.get("abbreviation", "")

    if abbreviation:
        if abbreviation in seen:
            duplicates.add(abbreviation)
        else:
            seen.add(abbreviation)


def detect_duplicate_ids(joined, duplicate_ids, seen_ids):
    if joined["_id"] in seen_ids:
        duplicate_ids.add(joined["_id"])
    else:
        seen_ids.add(joined["_id"])


def detect_duplicate_isolate_ids(joined, duplicate_isolate_ids):
    duplicates = set()

    isolate_ids = [i["id"] for i in joined["isolates"]]

    for isolate_id in isolate_ids:
        if isolate_ids.count(isolate_id) > 1:
            duplicates.add(isolate_id)

    if duplicates:
        duplicate_isolate_ids[joined["_id"]] = {
            "name": joined["name"],
            "duplicates": list(duplicates)
        }


def detect_duplicate_sequence_ids(joined, duplicate_sequence_ids, seen_sequence_ids):
    sequence_ids = virtool.otus.extract_sequence_ids(joined)

    # Add sequence ids that are duplicated within an OTU to the duplicate set.
    duplicate_sequence_ids.update({i for i in sequence_ids if sequence_ids.count(i) > 1})

    sequence_ids = set(sequence_ids)

    # Add sequence ids that have already been seen and are in the OTU.
    duplicate_sequence_ids.update(seen_sequence_ids & sequence_ids)

    # Add all sequences to seen list.
    seen_sequence_ids.update(sequence_ids)


def detect_duplicate_name(joined, duplicates, seen):
    lowered = joined["name"].lower()

    if joined["name"].lower() in seen:
        duplicates.add(joined["name"])
    else:
        seen.add(lowered)


def detect_duplicates(otus, strict=True):
    duplicate_abbreviations = set()
    duplicate_ids = set()
    duplicate_isolate_ids = dict()
    duplicate_names = set()
    duplicate_sequence_ids = set()

    seen_abbreviations = set()
    seen_ids = set()
    seen_names = set()
    seen_sequence_ids = set()

    for joined in otus:
        detect_duplicate_abbreviation(
            joined,
            duplicate_abbreviations,
            seen_abbreviations
        )

        detect_duplicate_name(
            joined,
            duplicate_names,
            seen_names
        )

        if strict:
            detect_duplicate_ids(
                joined,
                duplicate_ids,
                seen_ids,
            )

            detect_duplicate_isolate_ids(
                joined,
                duplicate_isolate_ids
            )

            detect_duplicate_sequence_ids(
                joined,
                duplicate_sequence_ids,
                seen_sequence_ids
            )

    errors = list()

    if duplicate_abbreviations:
        errors.append({
            "id": "duplicate_abbreviations",
            "message": "Duplicate OTU abbreviations found",
            "duplicates": list(duplicate_abbreviations)
        })

    if duplicate_ids:
        errors.append({
            "id": "duplicate_ids",
            "message": "Duplicate OTU ids found",
            "duplicates": list(duplicate_ids)
        })

    if duplicate_isolate_ids:
        errors.append({
            "id": "duplicate_isolate_ids",
            "message": "Duplicate isolate ids found in some OTUs",
            "duplicates": duplicate_isolate_ids
        })

    if duplicate_names:
        errors.append({
            "id": "duplicate_names",
            "message": "Duplicate OTU names found",
            "duplicates": list(duplicate_names)
        })

    if duplicate_sequence_ids:
        errors.append({
            "id": "duplicate_sequence_ids",
            "message": "Duplicate sequence ids found",
            "duplicates": duplicate_sequence_ids
        })

    return errors


def get_import_schema(require_meta=True):
    return {
        "data_type": {
            "type": "string",
            "required": require_meta
        },
        "organism": {
            "type": "string",
            "required": require_meta
        },
        "otus": {
            "type": "list",
            "required": True
        }
    }


def get_isolate_schema(require_id):
    return {
        "id": {
            "type": "string",
            "required": require_id
        },
        "source_type": {
            "type": "string",
            "required": True
        },
        "source_name": {
            "type": "string",
            "required": True
        },
        "default": {
            "type": "boolean",
            "required": True
        },
        "sequences": {
            "type": "list",
            "required": True
        }
    }


def get_otu_schema(require_id):
    return {
        "_id": {
            "type": "string",
            "required": require_id
        },
        "abbreviation": {
            "type": "string"
        },
        "name": {
            "type": "string",
            "required": True
        },
        "isolates": {
            "type": "list",
            "required": True
        }
    }


def get_owner_user(user_id):
    return {
        "id": user_id,
        "build": True,
        "modify": True,
        "modify_otu": True,
        "remove": True
    }


def get_sequence_schema(require_id):
    return {
        "_id": {
            "type": "string",
            "required": require_id
        },
        "accession": {
            "type": "string",
            "required": True
        },
        "definition": {
            "type": "string",
            "required": True
        },
        "sequence": {
            "type": "string",
            "required": True
        }
    }


def load_reference_file(path):
    """
    Load a list of merged otus documents from a file associated with a Virtool reference file.

    :param path: the path to the otus.json.gz file
    :type path: str

    :return: the otus data to import
    :rtype: dict

    """
    with open(path, "rb") as handle:
        with gzip.open(handle, "rt") as gzip_file:
            return json.load(gzip_file)


def validate_otu(otu, strict):
    report = {
        "otu": None,
        "isolates": dict(),
        "sequences": dict()
    }

    otu_validator = Validator(get_otu_schema(strict), allow_unknown=True)

    if not otu_validator.validate(otu):
        report["otu"] = otu_validator.errors

    report["isolates"] = dict()

    if "isolates" in otu:
        isolate_validator = Validator(get_isolate_schema(strict), allow_unknown=True)
        sequence_validator = Validator(get_sequence_schema(strict), allow_unknown=True)

        for isolate in otu["isolates"]:
            if not isolate_validator.validate(isolate):
                report["isolates"][isolate["id"]] = isolate_validator.errors

            if "sequences" in isolate:
                for sequence in isolate["sequences"]:
                    if not sequence_validator.validate(sequence):
                        report["sequences"][sequence["_id"]] = isolate_validator.errors

    if any(value for value in report.values()):
        return report

File: packages/contracts/src/references.ts
import { z } from "zod";
import type { SearchResult } from "./search";
import type { Task } from "./tasks";
import type { UserNested } from "./users";

/** The three per-reference rights a member (user or group) can be granted. */
export type ReferenceRights = {
	build: boolean;
	modify: boolean;
	modifyOtu: boolean;
};

/** The name of a single reference right. */
export type ReferenceRight = keyof ReferenceRights;

/** A user granted rights on a reference. */
export type ReferenceUser = ReferenceRights & {
	id: number;
	handle: string;
	createdAt: Date;
};

/** A group granted rights on a reference. */
export type ReferenceGroup = ReferenceRights & {
	id: number;
	name: string;
	createdAt: Date;
};

/** The reference a clone was created from, reduced to id and name. */
export type ReferenceClonedFrom = { id: number; name: string };

/** A contributor to a reference's history, with their change count. */
export type ReferenceContributor = UserNested & { count: number };

/** The most recent ready build (index) of a reference. */
export type ReferenceBuild = {
	id: number;
	version: number;
	createdAt: Date;
	user: UserNested;
};

/** The upload a reference was imported from, with its uploader. */
export type ReferenceImportedFrom = {
	id: number;
	name: string;
	createdAt: Date | null;
	size: number | null;
	user: UserNested | null;
};

/** A reference as it appears in a search-result list. */
export type ReferenceMinimal = {
	id: number;
	dataType: string;
	name: string;
	archived: boolean;
	clonedFrom: ReferenceClonedFrom | null;
	createdAt: Date;
	importedFrom: ReferenceImportedFrom | null;
	latestBuild: ReferenceBuild | null;
	organism: string;
	otuCount: number;
	task: Task | null;
	user: UserNested | null;
};

/** A full reference, as returned by the detail endpoint. */
export type Reference = ReferenceMinimal & {
	contributors: ReferenceContributor[];
	description: string;
	groups: ReferenceGroup[];
	restrictSourceTypes: boolean;
	sourceTypes: string[];
	users: ReferenceUser[];
};

/** A page of references. */
export type ReferenceSearchResult = SearchResult & {
	items: ReferenceMinimal[];
};

/**
 * Fields accepted when creating a reference. An empty name, description, or
 * organism is allowed; a clone fills the name in from its source. At most one
 * of `cloneFrom` and `importFrom` may be set — the server function enforces
 * that on top of this shape.
 */
export const ReferenceCreateRequest = z.object({
	name: z.string().trim().default(""),
	description: z.string().trim().default(""),
	organism: z.string().trim().default(""),
	cloneFrom: z.number().int().positive().optional(),
	importFrom: z.number().int().positive().optional(),
});

export type ReferenceCreateRequest = z.infer<typeof ReferenceCreateRequest>;

/** Fields accepted when updating a reference. Only those present are changed. */
export const ReferenceUpdateRequest = z.object({
	name: z.string().trim().min(1).optional(),
	description: z.string().optional(),
	organism: z.string().optional(),
	restrictSourceTypes: z.boolean().optional(),
	sourceTypes: z.array(z.string()).optional(),
});

export type ReferenceUpdateRequest = z.infer<typeof ReferenceUpdateRequest>;

/*
 * The shape of an uploaded reference file, and the port of Python's
 * `ReferenceSourceData` and the models under it.
 *
 * Two upload formats parse to this one shape: a gzipped JSON export, which
 * spells an OTU's and a sequence's id `_id`, and a `.v1.sqlite` snapshot, whose
 * reader yields the same documents keyed `id`. Python reconciles the two with a
 * pydantic alias plus `allow_population_by_field_name`; `withUnderscoreId`
 * below is that reconciliation, and it normalises onto `_id` because
 * `prepareOtuInsertion` reads the document Mongo-side.
 *
 * Every object is loose. A reference carries keys neither implementation reads
 * — `schema`, `taxid`, `remote` — and the OTU document is stored verbatim, so
 * stripping them would silently drop data on the way through.
 */

/**
 * Accept an `id` where the schema wants `_id`, as Python's alias does.
 *
 * Only fills `_id` in; an object that already has one is untouched, so a
 * document carrying both keeps whichever the JSON export wrote.
 */
function withUnderscoreId(value: unknown): unknown {
	if (typeof value !== "object" || value === null || Array.isArray(value)) {
		return value;
	}

	const record = value as Record<string, unknown>;

	if ("_id" in record || !("id" in record)) {
		return value;
	}

	const { id, ...rest } = record;

	return { _id: id, ...rest };
}

const referenceSourceSequenceSchema = z.preprocess(
	withUnderscoreId,
	z.looseObject({
		_id: z.string().min(1),
		accession: z.string().min(1),
		definition: z.string(),
		sequence: z.string().min(10),
	}),
);

const referenceSourceIsolateSchema = z.looseObject({
	// No alias here, and none in Python either: both readers spell an isolate's
	// id `id` already.
	id: z.string(),
	default: z.boolean(),
	source_type: z.string(),
	source_name: z.string(),
	sequences: z.array(referenceSourceSequenceSchema).min(1),
});

const referenceSourceOtuSchema = z.preprocess(
	withUnderscoreId,
	z.looseObject({
		_id: z.string(),
		abbreviation: z.string().default(""),
		name: z.string(),
		isolates: z.array(referenceSourceIsolateSchema).min(1),
	}),
);

/** One OTU of an uploaded reference, with its isolates and their sequences. */
export type ReferenceSourceOtu = z.infer<typeof referenceSourceOtuSchema>;

/**
 * The most duplicates of one kind named before the message is truncated.
 *
 * A file whose ids are all identical would otherwise put one entry per OTU into
 * `tasks.error`, which is read back into a popover.
 */
const MAX_REPORTED_DUPLICATES = 5;

/**
 * Every duplicate check Python runs, in one pass over the OTUs.
 *
 * Python spreads these across four `@validator("otus")` functions, which stop
 * at the first that raises, so a file with two problems is fixed and
 * re-uploaded only to fail on the next. Running them together reports all four
 * at once; the predicates themselves are unchanged, quirks included — the name
 * check compares case-insensitively but reports the name as written, isolate
 * ids collide only within their own OTU, and sequence ids collide across the
 * whole file.
 */
function checkForDuplicates(
	data: { otus: ReferenceSourceOtu[] },
	ctx: z.RefinementCtx,
): void {
	const seenNames = new Set<string>();
	const duplicateNames = new Set<string>();

	const seenOtuIds = new Set<string>();
	const duplicateOtuIds = new Set<string>();

	const seenSequenceIds = new Set<string>();
	const duplicateSequenceIds = new Set<string>();

	const otusWithDuplicateIsolates: string[] = [];

	for (const otu of data.otus) {
		const lowered = otu.name.toLowerCase();

		if (seenNames.has(lowered)) {
			duplicateNames.add(otu.name);
		} else {
			seenNames.add(lowered);
		}

		if (seenOtuIds.has(otu._id)) {
			duplicateOtuIds.add(otu._id);
		} else {
			seenOtuIds.add(otu._id);
		}

		const isolateIds = new Set<string>();
		let hasDuplicateIsolate = false;

		for (const isolate of otu.isolates) {
			if (isolateIds.has(isolate.id)) {
				hasDuplicateIsolate = true;
			}

			isolateIds.add(isolate.id);

			for (const sequence of isolate.sequences) {
				if (seenSequenceIds.has(sequence._id)) {
					duplicateSequenceIds.add(sequence._id);
				} else {
					seenSequenceIds.add(sequence._id);
				}
			}
		}

		if (hasDuplicateIsolate) {
			otusWithDuplicateIsolates.push(`${otu.name} (${otu._id})`);
		}
	}

	const problems: Array<[string, Iterable<string>]> = [
		["Duplicate OTU names", duplicateNames],
		["Duplicate OTU ids", duplicateOtuIds],
		["OTUs with duplicate isolate ids", otusWithDuplicateIsolates],
		["Duplicate sequence ids", duplicateSequenceIds],
	];

	for (const [label, found] of problems) {
		const items = [...found];

		if (items.length > 0) {
			ctx.addIssue({
				code: "custom",
				path: ["otus"],
				message: `${label}: ${summarize(items)}`,
			});
		}
	}
}

/** Name the first few of `items`, counting off however many are left. */
function summarize(items: string[]): string {
	const shown = items.slice(0, MAX_REPORTED_DUPLICATES).join(", ");

	return items.length > MAX_REPORTED_DUPLICATES
		? `${shown} and ${items.length - MAX_REPORTED_DUPLICATES} more`
		: shown;
}

export const ReferenceSourceDataSchema = z
	.object({
		/*
		 * Python's `ReferenceDataType` has one member, so a barcode reference is
		 * rejected there and must be rejected here too — both runners claim these
		 * tasks until the cutover, and the looser side would accept a file the
		 * other refuses.
		 */
		data_type: z.enum(["genome"]).default("genome"),
		organism: z.string().default("Unknown"),
		otus: z.array(referenceSourceOtuSchema).min(1),
	})
	.superRefine(checkForDuplicates);

/** A parsed reference upload, ready to be written to a reference. */
export type ReferenceSourceData = z.infer<typeof ReferenceSourceDataSchema>;
